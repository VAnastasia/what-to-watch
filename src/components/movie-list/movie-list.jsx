import React, {PureComponent} from "react";
import propTypes from "prop-types";
import {connect} from 'react-redux';
import MovieCard from "../movie-card/movie-card.jsx";
import {GENRE_DEFAULT} from "../../const";

class MovieList extends PureComponent {
  constructor(props) {
    super(props);
    this.handleCardHover = this.handleCardHover.bind(this);
    this.handleCardOut = this.handleCardOut.bind(this);
    this.timer = null;

    this.state = {
      activeCard: -1,
    };
  }

  handleCardHover(id) {
    return (evt) => {
      evt.preventDefault();
      this.timer = setTimeout(() => {
        this.setState({
          activeCard: id,
        });
      }, 1000);
    };
  }

  handleCardOut() {
    clearTimeout(this.timer);
    this.setState({
      activeCard: -1,
    });
  }

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  getFiltredMovies(movies, activeGenre) {
    if (activeGenre !== GENRE_DEFAULT) {
      return movies.slice().filter((movie) => movie.genre === activeGenre);
    }
    return movies;
  }

  render() {
    const {movies, activeGenre} = this.props;
    const {activeCard} = this.state;
    const films = this.getFiltredMovies(movies, activeGenre);

    return (
      <div className="catalog__movies-list">
        {films.map((film, index) => {
          return (
            <MovieCard
              key={index}
              film={film}
              onHover={this.handleCardHover}
              onLeave={this.handleCardOut}
              activeCard={activeCard}
            />
          );
        })}
      </div>
    );
  }
}

MovieList.propTypes = {
  movies: propTypes.arrayOf(
      propTypes.object.isRequired
  ).isRequired,
  activeGenre: propTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  movies: state.movies,
  activeGenre: state.genre
});

export {MovieList};
export default connect(mapStateToProps)(MovieList);
