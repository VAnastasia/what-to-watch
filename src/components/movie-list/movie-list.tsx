import * as React from "react";
import MovieCard from "../movie-card/movie-card";
import withActiveCard from "../../hocs/with-active-card/with-active-card";
import {MovieTypes} from "../../types";

interface Props {
  movies: MovieTypes[];
  loadComments: () => void;
}

const MovieCardWrapped = withActiveCard(MovieCard);

const MovieList: React.FunctionComponent<Props> = (props: Props) => {
  const {movies, loadComments} = props;
  return (
    <div className="catalog__movies-list">
      {movies.map((film) => {
        return (
          <MovieCardWrapped
            key={film.id}
            film={film}
            loadComments={loadComments}
          />
        );
      })}
    </div>
  );
};

export default MovieList;
