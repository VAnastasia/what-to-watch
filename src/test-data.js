const promoFilm = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  year: 2014
};

const films = [
  {
    id: 1,
    title: `Fantastic Beasts: The Crimes of Grindelwald`,
    image: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  {
    id: 2,
    title: `Bohemian Rhapsody`,
    image: `img/bohemian-rhapsody.jpg`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  {
    id: 3,
    title: `Macbeth`,
    image: `img/macbeth.jpg`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  {
    id: 4,
    title: `Aviator`,
    image: `img/aviator.jpg`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  },
  {
    id: 5,
    title: `We need to talk about Kevin`,
    image: `img/we-need-to-talk-about-kevin.jpg`,
    video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  }
];

const film = {
  id: 1,
  title: `The Grand Budapest Hotel`,
  image: `img/bg-the-grand-budapest-hotel.jpg`,
  video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
  poster: `img/the-grand-budapest-hotel-poster.jpg`,
  genre: `drama`,
  year: `2014`,
  rating: 8.9,
  ratingCount: 240,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege. Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  director: `Wes Andreson`,
  actors: [
    `Bill Murray`,
    `Edward Norton`,
    `Jude Law`,
    `Willem Dafoe`
  ]
};

export {promoFilm, films, film};
