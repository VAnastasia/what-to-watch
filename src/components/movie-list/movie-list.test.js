import React from "react";
import renderer from "react-test-renderer";
import {MovieList} from "./movie-list.jsx";
import {films} from "../../test-data";

it(`Should MovieList render correctly`, () => {
  const tree = renderer
    .create(<MovieList
      movies={films}
      activeGenre={`All genres`}
    />,
    {
      createNodeMock: () => {
        return {};
      }
    })
    .toJSON();

  expect(tree).toMatchSnapshot();
});
