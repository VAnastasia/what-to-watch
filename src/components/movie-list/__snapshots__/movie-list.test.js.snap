// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should MovieList render correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        alt=""
        height="175"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="/"
      />
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        alt=""
        height="175"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="/"
      />
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        alt=""
        height="175"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="/"
      />
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        alt=""
        height="175"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="/"
      />
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    onMouseLeave={[Function]}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <video
        alt=""
        height="175"
        width="280"
      />
    </div>
    <h3
      className="small-movie-card__title"
    >
      <a
        className="small-movie-card__link"
        href="/"
      />
    </h3>
  </article>
</div>
`;
